### PR TITLE
Include support for nested Children used for getBackiStorageCollection

### PR DESCRIPTION
**Pull request by inode64**
Include phpunit test too

### CODE REVIEW

File: src/YunInternet/Libvirt/XMLImplement/SimpleXMLImplement.php [hunk 1]
```diff
@@ -72,8 +72,13 @@ public function getChildren($name, $filter = null, $wrapper = null): array
             };
         }
 
+        return $this->GetCollection($name, $filter, $wrapper, $nested, $this->simpleXMLElement);
+    }
+
+    private function GetCollection(string $name, $filter, object $wrapper, bool $nested, $xml)
```
Comment (yzslab) [inline src/YunInternet/Libvirt/XMLImplement/SimpleXMLImplement.php:78]: Please use lowerCamelCase style method name. Only begin with upper case when it's a abbreviation, like IOTune, which is the abbreviation of "Input Output Tune".
You don't have to invoke this function recursively, there is an method SimpleXMLElement::xpath(), which can simplify your code, like this:
```
...
$path = ".//" . $name;
foreach ($this->simpleXMLElement->xpath($path) as $child) {
    ...
}
...
```
$this->simpleXMLElement->{$name} can be replaced with XPath too: 
```
$path = "./" . $name
```

File: src/YunInternet/Libvirt/Configuration/StorageVolume/BackingStore.php [hunk 2]
```diff
@@ -13,7 +13,7 @@
 use YunInternet\Libvirt\XMLImplement\SingletonChild;
 
 /**
- * Class BackingStore
+ * Class backingStore
```
Comment (yzslab) [inline src/YunInternet/Libvirt/Configuration/StorageVolume/BackingStore.php:16]: Class name should be CamelCase.

**Review by yzslab — CHANGES_REQUESTED** [review]

**Comment by inode64** [comment]
Fix all

File: src/YunInternet/Libvirt/XMLImplement/SimpleXMLImplement.php (hunk 1 above)
Comment (inode64) [reply src/YunInternet/Libvirt/XMLImplement/SimpleXMLImplement.php]: Thank for your suggestion

File: src/YunInternet/Libvirt/Configuration/Domain/Device/Disk.php [hunk 3]
```diff
@@ -168,4 +168,13 @@ public function hasBacking(): bool
...
         return is_null($this->findChild("backingStore")) === false && $this->backingStore()->isActive();
     }
+
+    /**
+     * @param callable|null $filter A callable accept a BackingStore as parameter
+     * @return array BackingStore[]
+     */
+    public function getBackiStorageCollection($filter = null): array
+    {
+        return $this->getChildren("backingStore", $filter, BackingStore::class, true);
```
Comment (yzslab) [inline src/YunInternet/Libvirt/Configuration/Domain/Device/Disk.php:178]: Do you want to add a default filter here? As you known, there are some situations have an backingStore without type attribute (not active).

Comment (inode64) [reply src/YunInternet/Libvirt/Configuration/Domain/Device/Disk.php:178]: I don't understand what you mean, can you give me some example of filters?

Comment (yzslab) [reply src/YunInternet/Libvirt/Configuration/Domain/Device/Disk.php:178]: > I don't understand what you mean, can you give me some example of filters?

https://github.com/yzslab/php-libvirt-client/pull/12/files#diff-7481b12aee3e3e747e5273a6c7359bb4e5f41df5082b1c4a037ce042c517ebb9R400 
There are three backingStore element in the XML, but only two of them are active:
```
<backingStore type='...'>
    <backingStore type='...'>
        <backingStore/>
    </backingStore>
</backingStore>
```
I think you should add a default filter here to exclude the not active one.

Comment (inode64) [reply src/YunInternet/Libvirt/Configuration/Domain/Device/Disk.php:178]: I'm sorry , I have tried, but there was no way, I had never used callable functions

Comment (yzslab) [reply src/YunInternet/Libvirt/Configuration/Domain/Device/Disk.php:183]: > I'm sorry , I have tried, but there was no way, I had never used callable functions

I made some changes, tell me if there are any questions or recommendations.

Comment (yzslab) [reply src/YunInternet/Libvirt/Configuration/Domain/Device/Disk.php:183]: https://github.com/yzslab/php-libvirt-client/pull/12/commits/2bc71a20c96cc718b401416089b5cbf4b5e79200

**Comment by inode64** [comment]
I fix all!
Thanks

**Comment by yzslab** [comment]
Merged, thanks.

**Pull request closed without merging**